core/roboconf-dm-monitoring/src/main/java/net/roboconf/dm/monitoring/internal/MonitoredApplication.java

```diff
@@ -0,0 +1,228 @@
+				instanceContext.exports.add( export );
+			}
+
+			// Instance data.
+			for (final Map.Entry<String, String> dataEntry : instance.data.entrySet()) {
+				final VariableContextBean data = new VariableContextBean();
+				data.name = dataEntry.getKey();
+				data.value = dataEntry.getValue();
+				instanceContext.data.add( data );
+			}
+
+			// Finally add the context to the maps.
+			instancesByPath.put( path, instanceContext );
+			context.instances.add( instanceContext );
+			for (final String typeName : instanceContext.types) {
+				Set<InstanceContextBean> instancesOfType = context.instancesByType.get( typeName );
+				if (instancesOfType == null) {
+					instancesOfType = new LinkedHashSet<InstanceContextBean>();
+					context.instancesByType.put( typeName, instancesOfType );
+				}
+				instancesOfType.add( instanceContext );
+			}
+		}
+
+		// Once the contexts of all the instances have been created, we can add their relationships, which are:
+		// - the parent instance, or null for a root instance.
+		// - the children instances
+		// - the instances from which variables are imported.
+		for (final Instance instance : allInstances) {
+			final InstanceContextBean instanceContext = instancesByPath.get( computeInstancePath( instance ) );
+
+			final Instance parent = instance.getParent();
+			if (parent != null) {
+				instanceContext.parent = instancesByPath.get( computeInstancePath( parent ) );
+			}
+
+			final Collection<Instance> children = instance.getChildren();
+			if (!children.isEmpty()) {
+				for (final Instance child : children) {
+					instanceContext.children.add( instancesByPath.get( computeInstancePath( child ) ) );
+				}
+			}
+
+			for (final Map.Entry<String, Collection<Import>> importedByType : instance.getImports().entrySet()) {
+				for (final Import importedVariable : importedByType.getValue()) {
+					final ImportContextBean importContext = new ImportContextBean();
+					importContext.component = importedVariable.getComponentName();
+					importContext.instance = instancesByPath.get( importedVariable.getInstancePath() );
+					for (Map.Entry<String, String> entry : importedVariable.getExportedVars().entrySet()) {
+						final VariableContextBean variable = new VariableContextBean();
+						variable.name = entry.getKey();
+						variable.value = entry.getValue();
+						importContext.variables.add( variable );
+					}
+					instanceContext.imports.add( importContext );
+				}
+			}
+		}
+
+		return context;
+	}
+
+	/**
+	 * Find all the components and the facets extended by the given component.
+	 * <p>
+	 * The component itself is included in the returned set.
+	 * </p>
+	 *
+	 * @param baseComponent the component whose ancestors must be returned.
+	 * @return the names of all the ancestor components and facets of the given component.
+	 */
+	private static Set<String> findExtendedTypes( final Component baseComponent ) {
+		final LinkedList<AbstractType> types = new LinkedList<AbstractType>();
```

Comment:
You can do shorter.
ComponentHelpers.findAllExtendedComponents( baseComponent );
ComponentHelpers.findAllFacets( baseComponents );